source/index.ts

```diff
@@ -0,0 +1,317 @@
+					this.timeoutId = setTimeout(
+						() => {
+							this.onResumeInterval();
+						},
+						delay
+					);
+				}
+
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	tryToStartAnother(): boolean {
+		if (this.queue.size === 0) {
+			// We can clear the interval ("pause")
+			// Because we can redo it later ("resume")
+			if (this.intervalId) {
+				clearInterval(this.intervalId);
+			}
+			this.intervalId = undefined;
+
+			this.resolvePromises();
+
+			return false;
+		}
+
+		if (!this.paused) {
+			const canInitializeInterval = !this.intervalPaused();
+			if (this.doesIntervalAllowAnother && this.doesConcurrentAllowAnother) {
+				this.emit('active');
+
+				// tslint:disable-next-line:no-non-null-assertion no-floating-promises
+				this.queue.dequeue()!();
+				if (canInitializeInterval) {
+					this.initializeIntervalIfNeeded();
+				}
+
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	initializeIntervalIfNeeded(): void {
+		if (this.isIntervalIgnored || this.intervalId !== undefined) {
+			return;
+		}
+
+		this.intervalId = setInterval(
+			() => {
+				this.onInterval();
+			},
+			this.interval
+		);
+		this.intervalEnd = Date.now() + this.interval;
+	}
+
+	onInterval(): void {
+		if (this.intervalCount === 0 && this.pendingCount === 0 && this.intervalId) {
+			clearInterval(this.intervalId);
+			this.intervalId = undefined;
+		}
+
+		this.intervalCount = (this.carryoverConcurrencyCount) ? this.pendingCount : 0;
+		// tslint:disable-next-line:no-empty
+		while (this.tryToStartAnother()) {}
+	}
+
+	/**
+	 * Adds a sync or async task to the queue. Always returns a promise.
+	 */
+	async add<TaskResultType>(fn: Task<TaskResultType>, options?: EnqueueOptionsType): Promise<TaskResultType> {
+		return new Promise<TaskResultType>((resolve, reject) => {
+			const run = async () => {
+				this.pendingCount++;
+				this.intervalCount++;
+
+				try {
+					// tslint:disable-next-line:await-promise
+					resolve(await fn());
+				} catch (error) {
+					reject(error);
+				}
+
+				this.next();
+			};
+
+			this.queue.enqueue(run, options);
+			this.tryToStartAnother();
+		});
+	}
+
+	/**
+	 * Same as `.add()`, but accepts an array of sync or async functions.
+	 * @returns A promise that resolves when all functions are resolved.
+	 */
+	async addAll<TaskResultsType>(fns: Task<TaskResultsType>[], options?: EnqueueOptionsType): Promise<TaskResultsType[]> {
+		return Promise.all(fns.map(fn => this.add(fn, options)));
+	}
+
+	/**
+	 * Start (or resume) executing enqueued tasks within concurrency limit. No need to call this if queue is not paused (via `options.autoStart = false` or by `.pause()` method.)
+	 */
+	start(): void {
+		if (!this.paused) {
+			return;
+		}
+
+		this.paused = false;
+		// tslint:disable-next-line:no-empty
+		while (this.tryToStartAnother()) {}
+	}
+
+	/**
+	 * Put queue execution on hold.
+	 */
+	pause(): void {
+		this.paused = true;
+	}
+
+	/**
+	 * Clear the queue.
+	 */
```

Comment:
I use this style for comments:

```ts
/**
Clear the queue.
*/
```

Reply:
your tslint config was complaining for that style, now eslint doesn't care. Did you configure https://eslint.org/docs/rules/multiline-comment-style?

I'll change them manually for now though 😇